### PR TITLE
feature: migrate to gestaltv7

### PR DESCRIPTION
**Pull request by pollend**
Here is the initial migration to gestalt-v7.

## Changes

- update gestalt package
  - org.terasology.gestalt.assets.management.AssetManager
  - org.terasology.gestalt.naming.Name
  - org.terasology.gestalt.assets.AssetType;
  - org.terasology.gestalt.assets.management.AssetManager;
  - org.terasology.gestalt.assets.module.ModuleAwareAssetTypeManager;
  - org.terasology.gestalt.assets.module.ModuleAwareAssetTypeManagerImpl;
  - org.terasology.gestalt.module.ModuleEnvironment;
  - org.terasology.gestalt.module.ModuleRegistry;
  - org.terasology.gestalt.module.dependencyresolution.DependencyResolver;
  - org.terasology.gestalt.module.dependencyresolution.ResolutionResult;

## Issues

- [ ] something jmh-related preventing running from IntelliJ?
- [ ] tidy use of `google()` repository
- [x] #4610 Crash on F11 (DebugControlSystem failure to get MouseDevice) 
- [ ] migrate the modules: many have branches under the name `chore/gestalt-v7-migration`

----
## Module Changes

- [ ] https://github.com/Terasology/CoreRendering/pull/36

### CODE REVIEW

**Comment by pollend** [comment]
any clue why this isn't working

```
Terasology/engine/src/main/java/org/terasology/engine/core/module/ModuleManagerImpl.java:105:17
java: cannot access com.github.zafarkhaja.semver.Version
  class file for com.github.zafarkhaja.semver.Version not found
```

**Comment by keturn** [comment]
Adding it to [terasology-repositories in build-logic](https://github.com/MovingBlocks/Terasology/blob/cbc07339ea0fdbfaf1cfe31beee8b0b44a25834e/build-logic/src/main/kotlin/terasology-repositories.gradle.kts) would normally be sufficient for adding a repository, but since it's a dependency of gestalt and build-logic _uses_ gestalt itself, I think it'll also need to be in `buildscript { repository { } }` sections. That includes the root build.gradle as well as the templates/build.gradle for modules.

(Yeah, I'm starting to doubt the wisdom of making build-logic depend on gestalt. Or wonder if there's some way for build-logic to be responsible for its own repositories so it doesn't leak out to the other `buildscript` sections.)

**Comment by keturn** [comment]
I was just going to go look at ModuleManager and discovered that it's an interface again, undoing #4539 (which undid #1665).

did you discover a reason for that to be an interface that we missed when I submitted #4539?

File: engine-tests/src/test/java/org/terasology/documentation/BindingScraper.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
 import com.google.common.collect.ArrayListMultimap;
 import com.google.common.collect.Multimap;
 import org.terasology.engine.core.module.ModuleManager;
+import org.terasology.engine.core.module.ModuleManagerImpl;
```
Comment (keturn) [inline engine-tests/src/test/java/org/terasology/documentation/BindingScraper.java:21]: why are there a bunch of these changes adding an import but not anything that would use that import?

File: engine/build.gradle [hunk 2]
```diff
@@ -105,15 +110,19 @@ dependencies {
     implementation group: 'net.logstash.logback', name: 'logstash-logback-encoder', version: '4.10'
 
     // Our developed libs
-    api group: 'org.terasology', name: 'gestalt-module', version: '5.1.5'
-    api group: 'org.terasology', name: 'gestalt-util', version: '5.1.5'
-    api group: 'org.terasology', name: 'gestalt-asset-core', version: '5.1.5'
+    api group: 'org.terasology.gestalt', name: 'gestalt-asset-core', version: '7.0.6-SNAPSHOT'
+    api group: 'org.terasology.gestalt', name: 'gestalt-entity-system', version: '7.0.6-SNAPSHOT'
```
Comment (keturn) [inline engine/build.gradle:114]: including entity-system?

**Comment by pollend** [comment]
> I was just going to go look at ModuleManager and discovered that it's an interface again, undoing #4539 (which undid #1665).
> 
> did you discover a reason for that to be an interface that we missed when I submitted #4539?

I was referencing @DarkWeird older PR. I'll undo that change.

**Comment by pollend** [comment]
I'm having some strange rendering problems where you only see the last dropped item. not sure what is causing it and how to resolve the problem.

**Comment by DarkWeird** [comment]
> I'm having some strange rendering problems where you only see the last dropped item. not sure what is causing it and how to resolve the problem.

Dropped items has same `LocationComponent` because using default `CopyStrategy` because `nui-reflect` haven't reflections.cache (and Reflections didn't fetch nui-reflect at runtime)

Probably it can be resolved with java 8 (reflections have several glitches at java 11... classloadersss...)
Or provide `reflections.cache from `nui-reflect`

**Comment by keturn** [comment]
If this is using https://github.com/MovingBlocks/TeraNUI/pull/46, does that mean we _don't_ need to do the change to the NUI namespaces in https://github.com/MovingBlocks/TeraNUI/pull/27 at the same time?

**Comment by DarkWeird** [comment]
> If this is using https://github.com/MovingBlocks/TeraNUI/pull/46, does that mean we _don't_ need to do the change to the NUI namespaces in https://github.com/MovingBlocks/TeraNUI/pull/27 at the same time?
> 
> edit: oh, wait, that's what DW's comment just prior to this one was about, things failing because a nui package is not part of `org.terasology.nui`.

I think use BSA's PR firsts, then provide support in nui-gestalv7 module is better then syntetic module

**Comment by keturn** [comment]
Sounds like you've been learning about what works and what doesn't.

We are going to end up with a PR that changes a zillion import statements; with changing the packages of things like `Name`, `ResourceUrn`, and `API`, that's unavoidable. Are there any pieces of this we can pull out and do in more easy-to-review PRs?

The thing with `engine/src/jmh` seems like it's probably one, since it moves a bunch of stuff but I'm not really sure what that's about.

Are there other fixes that would also apply to the current code even without the gestalt change? e.g. the camera position?

**Comment by keturn** [comment]
I haven't tried it out yet, but I strongly suspect the thing you described with `jmh` being pulled in to the reflections is from this:

https://github.com/MovingBlocks/Terasology/blob/21d13a3aeccaf649042e024c23d90bedeb1a8e9b/engine/build.gradle#L130-L134

note how the comment says "all sourceSets." Which might be useful if we had parts of engine written in Kotlin or Scala, but is probably not helping us as it is.

And yes, I'm still hoping to roll back the jmh part of this PR. Not just for size; also because I think in general we want to move in the direction of "move things from engine-test/src/test to engine/src/test" instead of "move things from engine to engine-test." Admittedly I am really not at all clear on where the `jmh` sources fall in that or how they are used.

**Comment by pollend** [comment]
> I haven't tried it out yet, but I strongly suspect the thing you described with `jmh` being pulled in to the reflections is from this:
> 
> https://github.com/MovingBlocks/Terasology/blob/21d13a3aeccaf649042e024c23d90bedeb1a8e9b/engine/build.gradle#L130-L134
> 
> note how the comment says "all sourceSets." Which might be useful if we had parts of engine written in Kotlin or Scala, but is probably not helping us as it is.
> 
> And yes, I'm still hoping to roll back the jmh part of this PR. Not just for size; also because I think in general we want to move in the direction of "move things from engine-test/src/test to engine/src/test" instead of "move things from engine to engine-test." Admittedly I am really not at all clear on where the `jmh` sources fall in that or how they are used.

yep, go ahead and revert the jmh part of this PR.

**Comment by keturn** [comment]
Getting two NPEs after I save a screen shot. One of them is coming from NUISkinEditorScreen, which is very puzzling given that I don't have that screen open nor have I used it this run at all:

```
[main] ERROR o.t.engine.core.TerasologyEngine - Uncaught exception, attempting clean game shutdown
com.google.common.base.VerifyException: Module "assets" not found in current module environment.
	at com.google.common.base.Verify.verify(Verify.java:124)
	at com.google.common.base.Verify.verifyNotNull(Verify.java:500)
	at org.terasology.engine.rendering.nui.editor.layers.AbstractEditorScreen.getPath(AbstractEditorScreen.java:336)
	at org.terasology.engine.rendering.nui.editor.layers.NUISkinEditorScreen.setSelectedAssetPath(NUISkinEditorScreen.java:510)
	at org.terasology.engine.rendering.nui.editor.layers.AbstractEditorScreen.loadAutosave(AbstractEditorScreen.java:299)
	at org.terasology.engine.rendering.nui.editor.layers.AbstractEditorScreen.onDraw(AbstractEditorScreen.java:157)
	at org.terasology.nui.canvas.CanvasImpl.drawStyledWidget(CanvasImpl.java:442)
	at org.terasology.nui.canvas.CanvasImpl.drawWidget(CanvasImpl.java:428)
	at org.terasology.engine.rendering.nui.internal.NUIManagerInternal.render(NUIManagerInternal.java:570)
	at org.terasology.engine.core.modes.StateIngame.renderUserInterface(StateIngame.java:231)
	at org.terasology.engine.core.modes.StateIngame.render(StateIngame.java:215)
	at org.terasology.engine.core.subsystem.lwjgl.LwjglGraphics.postUpdate(LwjglGraphics.java:89)
	at org.terasology.engine.core.TerasologyEngine.tick(TerasologyEngine.java:497)
```

**Comment by keturn** [comment]
The other is from DebugControlSystem:

```
[main] ERROR o.t.e.e.e.internal.EventSystemImpl - Failed to invoke event
java.lang.NullPointerException: null
	at org.terasology.engine.logic.players.DebugControlSystem.onKeyDown(DebugControlSystem.java:133)
	at org.terasology.engine.logic.players.DebugControlSystemMethodAccess.invoke(Unknown Source)
	at org.terasology.engine.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:389)
	at org.terasology.engine.entitySystem.event.internal.EventSystemImpl.sendConsumableEvent(EventSystemImpl.java:268)
	at org.terasology.engine.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:247)
	at org.terasology.engine.core.bootstrap.eventSystem.AbstractEventSystemDecorator.send(AbstractEventSystemDecorator.java:67)
	at org.terasology.engine.network.NetworkEventSystemDecorator.send(NetworkEventSystemDecorator.java:54)
	at org.terasology.engine.core.bootstrap.eventSystem.AbstractEventSystemDecorator.send(AbstractEventSystemDecorator.java:67)
	at org.terasology.engine.recording.RecordingEventSystemDecorator.send(RecordingEventSystemDecorator.java:34)
	at org.terasology.engine.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:205)
	at org.terasology.engine.input.InputSystem.send(InputSystem.java:525)
	at org.terasology.engine.input.InputSystem.sendKeyEvent(InputSystem.java:454)
	at org.terasology.engine.input.InputSystem.processKeyboardInput(InputSystem.java:383)
	at org.terasology.engine.input.InputSystem.update(InputSystem.java:131)
	at org.terasology.engine.core.modes.StateIngame.handleInput(StateIngame.java:192)
	at org.terasology.engine.core.subsystem.lwjgl.LwjglInput.postUpdate(LwjglInput.java:39)
	at org.terasology.engine.core.TerasologyEngine.tick(TerasologyEngine.java:497)
	at org.terasology.engine.core.TerasologyEngine.mainLoop(TerasologyEngine.java:450)
	at org.terasology.engine.core.TerasologyEngine.runMain(TerasologyEngine.java:426)
	at org.terasology.engine.core.TerasologyEngine.run(TerasologyEngine.java:392)
	at org.terasology.engine.Terasology.main(Terasology.java:179)
```

[I really wish java or github or something in this toolchain would inline sources in tracebacks.]

We're not on Java 14 yet so we don't have useful NPE messages, but that line is https://github.com/MovingBlocks/Terasology/blob/7143a6c10bd09cd48a96d39538d2cabb99c6719b/engine/src/main/java/org/terasology/engine/logic/players/DebugControlSystem.java#L133

so I guess it's probably `mouseDevice` that is null???

**Comment by keturn** [comment]
* ERROR o.t.engine.registry.InjectionHelper - BlockCommands wanted WorldProviderCoreImpl injected but CoreRegistry has none.
* ERROR o.t.engine.registry.InjectionHelper - DebugControlSystem wanted MouseDevice injected but CoreRegistry has none.

**Comment by keturn** [comment]
I've made progress on the test failures. Still a handful left to work through.

A handful left in engine-test, that is. Haven't checked to see if MTE works at all or what the rest of modulespace looks like.

**Comment by keturn** [comment]
From DarkWeird on [#stabilization](https://discord.com/channels/270264625419911192/696830442065625118/827564437178679306):

> DS provides NUI as hardcoded method.  
> TS did same way at gestaltv7 prs.
>
> Nui as syntetic midule listes in module lists. Have all rights as module.  
> Nui as syntetic module uses gestaltv7's mechanism for registration package modules + handwritten code for adopting.
>
> Nuiv3 provides as normal third party library. Don't register as module.  
> It is fine. Imho

File: engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java [hunk 3]
```diff
@@ -110,27 +125,6 @@ private void loadModulesFromApplicationPath(PathManager pathManager) {
         scanner.scan(registry, paths);
     }
 
-    private Module loadEngineModule(List<Class<?>> classesOnClasspathsToAddToEngine) {
-        // FIXME: is `classes…toAddToEngine` gone? Did we ever use it in the first place?
```
Comment (keturn) [inline engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java]: Did you check for `classesOnClasspathsToAddToEngine` usage? I know it's not usually used, so maybe this is a non-issue, but I had a vague memory about it being used somewhere...

File: engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java [hunk 4]
```diff
@@ -242,19 +236,4 @@ public ModuleFactory getModuleFactory() {
         return moduleFactory;
     }
 
-//    private void enrichReflectionsWithSubsystems(Module engineModule) {
-//        Serializer serializer = new XmlSerializer();
```
Comment (keturn) [inline engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java]: why _doesn't_ the gestalt-v7 branch need an `enrichReflectionsWithSubsystems` method? v7 hasn't done away with Reflections yet…

File: engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java [hunk 5]
```diff
@@ -73,17 +73,32 @@ public ModuleManager(String masterServerAddress, List<Class<?>> classesOnClasspa
...
+        loadEnvironment(Sets.newHashSet(engineModule, nui, discordSubsystem), true);
         installManager = new ModuleInstallManager(this, masterServerAddress);
     }
 
+    private  void createDependency(Module parent, Module child) {
+        DependencyInfo dependencyInfo = new DependencyInfo();
+        dependencyInfo.setId(child.getId());
+        dependencyInfo.setMinVersion(child.getVersion());
+        dependencyInfo.setMaxVersion(child.getVersion().getNextPatchVersion());
+        parent.getMetadata().getDependencies().add(dependencyInfo);
```
Comment (keturn) [inline engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java:99]: 👍 makes sense to have a function for this.

TODO: refactor `ensureModulesDependOnEngine` method (just below this one) to use this.

FIXME: note https://github.com/MovingBlocks/Terasology/pull/4593/commits/20546f1853a8f5cce99fff8018700d81340181d3 — getNextPatchVersion may break for this purpose when base version is a SNAPSHOT.

File: engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java [hunk 6]
```diff
@@ -73,17 +73,32 @@ public ModuleManager(String masterServerAddress, List<Class<?>> classesOnClasspa
...
+        engineModule = moduleFactory.createPackageModule("org.terasology.engine");
+        Module nui = createSyntheticPackageModule("nui", "3.0.0", "org.terasology.nui");
+        Module discordSubsystem = moduleFactory.createPackageModule("org.terasology.subsystem.discordrpc");
 
-        loadModulesFromApplicationPath(pathManager);
+        createDependency(engineModule, nui);
+
+        registry.add(nui);
+        registry.add(engineModule);
+        registry.add(discordSubsystem);
```
Comment (keturn) [inline engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java:84]: hrmmm. Not sure about adding hardcoded references to subsystems in engine's ModuleManager.

The point of extracting subsystems—as I understand it—is to make it so the engine _doesn't_ always need to depend on them. i.e. a facade should be able to start an engine that doesn't have a DiscordRPC subsystem at all.

File: engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java [hunk 7]
```diff
@@ -73,17 +73,32 @@ public ModuleManager(String masterServerAddress, List<Class<?>> classesOnClasspa
...
 
+        loadModulesFromApplicationPath(pathManager);
         ensureModulesDependOnEngine();
 
         setupSandbox();
-        loadEnvironment(Sets.newHashSet(engineModule), true);
+        loadEnvironment(Sets.newHashSet(engineModule, nui, discordSubsystem), true);
```
Comment (keturn) [inline engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java:90]: TODO: We should switch this to the new `resolveAndLoadEnvironment` method. It'll do dependency resolution so we don't have to manually keep these lists in sync with every change.

**Review by keturn — COMMENTED** [review]
I think like we have some things to clarify about the fundamental the goals of subsystem extraction.

**Comment by keturn** [comment]
Build #25 had 2 failing tests: http://jenkins.terasology.io/teraorg/job/Terasology/job/engine/job/PR-4593/25/

it has now regressed to 43 failures.

and PlayerConfig-to-AutoConfig was **inculded** in build 25, so that wasn't the problem.

**Comment by pollend** [comment]
I decided to move the discord package since trying to register it to the environment is very difficult and harder to maintain if we introduce more subsystems. @keturn

**Comment by keturn** [comment]
I went back a few steps to see if I could reproduce the troubles you were having with DiscordRPC. and indeed I did! so I'm not sure how I had things working as well as I remember; maybe they worked _until_ I merged in the AutoConfig changes from develop, and I didn't re-check after that? That's my best guess.

I also discovered that we seem to have the modules using the JsonSerializer for Reflections but engine and subsystems are still on XML, which is sure to be a source of confusion.

**Comment by keturn** [comment]
I plan to experiment a little to see if we can work things without sacrificing the engine/subsystem separation.

The plan is something like:

1. Make sure engine's Reflections are actually being loaded.
2. try re-instituting `enrichReflectionsWithSubsystems`
3. and probably switch all the Reflections caches to the JSON serializer for consistency, even if it's not strictly required by the above

**Comment by keturn** [comment]
@pollend Which thing was it that called for the `engine-test/…/module.txt` file to move?

**Comment by pollend** [comment]
uuuh, I don't remember why 

> @pollend Which thing was it that called for the `engine-test/…/module.txt` file to move?

**Comment by jdrueckert** [comment]
[[GESTALTv7] migrate module & asset](https://trello.com/c/te4uIhaz/46-gestaltv7-migrate-module-asset)

**Comment by keturn** [comment]
superceeded by #4622

**Pull request closed without merging**